### PR TITLE
Simpler v2 cgroup interface

### PR DESCRIPTION
**Pull request by crosbymichael**
The resource type is changed in this but I would rather have a v2 native
Resource type and do a mapping from an OCI spec's resource to the v2.Resource
type.

ref #104 

Signed-off-by: Michael Crosby <crosbymichael@gmail.com>

### CODE REVIEW

**Comment by crosbymichael** [comment]
@AkihiroSuda and @Zyqsempai please take a look at this new approach.  it should make things much easier to manage and extend.

File: v2/controller.go [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+		r.Memory,
+		r.Pids,
+	}
+	for _, v := range values {
+		if v == nil {
+			continue
+		}
+		o = append(o, v.Values()...)
+	}
+	return o
```
Comment (Zyqsempai) [inline v2/controller.go]: Can you make the name of this variable more explicit?

File: v2/controller.go [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+	}
+	return ioutil.WriteFile(
+		filepath.Join(path, c.key),
+		data,
+		perm,
+	)
+}
+
+func writeCGStr(path string, strs []CGStr) error {
+	for _, o := range strs {
```
Comment (Zyqsempai) [inline v2/controller.go]: Same here, actually I can't get what `o` stands for;)

File: v2/controller.go [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+		if err := o.write(path, defaultFilePerm); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func NewController(mountpoint string, g string, resources *Resources) (*Controller, error) {
```
Comment (Zyqsempai) [inline v2/controller.go]: Same, since `g` not any more of the GroupPath type, it's not so obvious what it stands for.

File: v2/controller.go [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+	for _, v := range values {
+		if v == nil {
+			continue
+		}
+		o = append(o, v.Values()...)
+	}
+	return o
+}
+
+type CGStr struct {
```
Comment (Zyqsempai) [inline v2/controller.go]: Also, the naming of this struct raises some questions.

File: v2/controller.go [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+		if current == state {
+			return nil
+		}
+		time.Sleep(1 * time.Millisecond)
+	}
+}
+
+func (c *Controller) Stat(g string) (*statsv2.Metrics, error) {
+	var stats statsv2.Metrics
+	return &stats, nil
```
Comment (Zyqsempai) [inline v2/controller.go]: Sorry, I didn't get it, how do you plan to collect stats for different controllers?

File: v2/controller.go (hunk 1 above)
Comment (fuweid) [reply v2/controller.go]: it is ok to use short name here. `o` sounds like output.:P

https://github.com/golang/go/wiki/CodeReviewComments#variable-names

File: v2/controller.go [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+		return nil, err
+	}
+
+	path := filepath.Join(mountpoint, string(g))
+	if err := os.MkdirAll(path, defaultDirPerm); err != nil {
+		return nil, err
+	}
+	if err := writeCGStr(path, resources.Values()); err != nil {
+		// clean up cgroup dir on failure
+		os.Remove(path)
```
Comment (fuweid) [inline v2/controller.go]: should be remove folder just in case it is not empty dir.

File: v2/controller.go [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+	if strings.HasPrefix(name, "/") {
+		return nil, errors.New("name must be relative")
+	}
+	path := filepath.Join(c.path, name)
+	if err := os.MkdirAll(path, defaultDirPerm); err != nil {
+		return nil, err
+	}
+	if err := writeCGStr(path, resources.Values()); err != nil {
+		// clean up cgroup dir on failure
+		os.Remove(path)
```
Comment (fuweid) [inline v2/controller.go]: should be remove folder by `os.RemoveAll` just in case it is not empty dir.

File: v2/controller.go (hunk 5 above)
Comment (fuweid) [reply v2/controller.go]: cgroupv2 will list enabled controllers in `cgroup.controllers` file. I think it is easy to collect the data for the controllers.

File: v2/utils.go [hunk 8]
```diff
@@ -167,7 +148,7 @@ func parseCgroupFromReader(r io.Reader) (GroupPath, error) {
 		}
 		// text is like "0::/user.slice/user-1001.slice/session-1.scope"
 		if parts[0] == "0" && parts[1] == "" {
-			return GroupPath(parts[2]), nil
```
Comment (fuweid) [inline v2/utils.go]: still feel that it is not more explicit. the pattern is `0::PATH` and not need to split it into 3 parts😂

File: v2/controller.go (hunk 4 above)
Comment (AkihiroSuda) [reply v2/controller.go]: `Value`?

File: v2/controller.go [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+	}
+	return o
+}
+
+type CGStr struct {
+	key   string
+	value interface{}
+}
+
+func (c *CGStr) write(path string, perm os.FileMode) error {
```
Comment (AkihiroSuda) [inline v2/controller.go]: please add a code comment to clarify `path` is file path (including `/sys/fs/cgroup` prefix), not a group path?

File: v2/controller.go [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+		return nil, err
+	}
+	path := filepath.Join(mountpoint, string(g))
+	return &Controller{
+		unifiedMountpoint: mountpoint,
+		path:              path,
+	}, nil
+}
+
+type Controller struct {
```
Comment (AkihiroSuda) [inline v2/controller.go]: Manager?

File: v2/paths.go [hunk 11]
```diff
@@ -22,33 +22,26 @@ import (
...
-func PidGroupPath(pid int) (GroupPath, error) {
+func PidGroupPath(pid int) (string, error) {
 	p := fmt.Sprintf("/proc/%d/cgroup", pid)
 	return parseCgroupFile(p)
 }
 
 // VerifyGroupPath verifies the format of g.
 // VerifyGroupPath doesn't verify whether g actually exists on the system.
-func VerifyGroupPath(g GroupPath) error {
+func VerifyGroupPath(g string) error {
```
Comment (AkihiroSuda) [inline v2/paths.go]: Please either move `GroupPath` godoc here, or restore `GroupPath` type?

```
// GroupPath is a string that appears as the third field in /proc/PID/cgroup.
// e.g. "/user.slice/user-1001.slice/session-1.scope"
//
// GroupPath must not contain "/sys/fs/cgroup" prefix.
// GroupPath must be a absolute path starts with "/".
```

File: v2/cpu.go [hunk 12]
```diff
@@ -16,108 +16,23 @@
...
+	if r.Weight != nil {
+		o = append(o, CGStr{
+			key:   "cpu.weight",
+			value: *r.Weight,
+		})
 	}
-	if cpuShares := resources.CPU.Shares; cpuShares != nil {
-		// Converting cgroups configuration from v1 to v2
-		// more here https://github.com/containers/crun/blob/master/crun.1.md#cgroup-v2
-		convertedWeight := (1 + ((*cpuShares-2)*9999)/262142)
```
Comment (AkihiroSuda) [inline v2/cpu.go]: where is the new code for this?

File: v2/controller.go [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+func (c *Controller) AddProc(pid uint64) error {
+	v := CGStr{
+		key:   cgroupProcs,
+		value: pid,
+	}
+	return writeCGStr(c.path, []CGStr{v})
+}
+
+func (c *Controller) Delete() error {
+	return remove(c.path)
```
Comment (AkihiroSuda) [inline v2/controller.go]: can be just `os.Remove`?

**Comment by AkihiroSuda** [comment]
`cmd/cgroups-playground` needs to be updated

File: v2/controller.go (hunk 6 above)
Comment (crosbymichael) [reply v2/controller.go]: In cgroups, it's usually just a `rm`

File: v2/controller.go (hunk 5 above)
Comment (crosbymichael) [reply v2/controller.go]: As for the implementation, i'm still working on a simpler way to the files and parse them easily without a lot of switches.  For now, we can just omit stats right now.

File: v2/cpu.go (hunk 12 above)
Comment (crosbymichael) [reply v2/cpu.go]: i'm skipping this for now to make sure we have a simple API first, then we can port the spec -> resource type code over along with stats.  The old API is too complicated and has too much code.

File: v2/controller.go (hunk 13 above)
Comment (crosbymichael) [reply v2/controller.go]: I think we still need the remove all logic but i'm not sure if this is more robust in v2

File: v2/manager.go [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+
+type Manager struct {
+	unifiedMountpoint string
+	path              string
+}
+
+func (c *Manager) NewChild(name string, resources *Resources) (*Manager, error) {
+	if strings.HasPrefix(name, "/") {
+		return nil, errors.New("name must be relative")
+	}
```
Comment (AkihiroSuda) [inline v2/manager.go]: write `cgroup.subtree_control`?

File: v2/paths.go [hunk 15]
```diff
@@ -22,33 +22,26 @@ import (
...
 	s := string(g)
```
Comment (AkihiroSuda) [inline v2/paths.go]: can be removed

File: v2/paths.go [hunk 16]
```diff
@@ -22,33 +22,26 @@ import (
...
 // PidGroupPath will return the correct cgroup paths for an existing process running inside a cgroup
 // This is commonly used for the Load function to restore an existing container
-func PidGroupPath(pid int) (GroupPath, error) {
+func PidGroupPath(pid int) (string, error) {
 	p := fmt.Sprintf("/proc/%d/cgroup", pid)
 	return parseCgroupFile(p)
 }
 
 // VerifyGroupPath verifies the format of g.
 // VerifyGroupPath doesn't verify whether g actually exists on the system.
```
Comment (AkihiroSuda) [inline v2/paths.go]: I still think we need godoc to clarify what is the expected format.
Probably we should add an example string like `"/user.slice/user-1001.slice/session-1.scope"`

File: v2/memory.go [hunk 17]
```diff
@@ -132,11 +60,11 @@ func (m *memoryController) Stat(g GroupPath, stats *statsv2.Metrics) error {
 		if err := sc.Err(); err != nil {
 			return err
 		}
-		key, v, err := parseKV(sc.Text())
+		filename, v, err := parseKV(sc.Text())
```
Comment (Zyqsempai) [inline v2/memory.go]: It's not so clear, why `parseKV`(parse Key and value) function returns filename. I think `key` was way more clear.

**Comment by crosbymichael** [comment]
Ok, pushed a big update. This has stats implemented along with renaming the playground tool to `cgctl` with a couple methods to play with this.  

```bash
NAME:
   cgctl - cgroup v2 management tool

USAGE:
   cgctl [global options] command [command options] [arguments...]

VERSION:
   1

COMMANDS:
   new      create a new cgroup
   del      delete a cgroup
   list     list processes in a cgroup
   stat     stat a cgroup
   help, h  Shows a list of commands or help for one command

GLOBAL OPTIONS:
   --debug             enable debug output in the logs
   --mountpoint value  cgroup mountpoint (default: "/sys/fs/cgroup")
   --help, -h          show help
   --version, -v       print the version
```

```bash
> sudo ./cgctl new --enable test
> sudo ./cgctl list test
> sudo ./cgctl stat test
> sudo ./cgctl del test
```

File: v2/paths.go (hunk 15 above)
Comment (AkihiroSuda) [reply v2/paths.go]: I meant type casting can be removed. The function should be kept?

File: cmd/cgctl/main.go [hunk 18]
```diff
@@ -0,0 +1,141 @@
...
+		c, err := v2.NewManager(clix.GlobalString("mountpoint"), path, nil)
+		if err != nil {
+			return err
+		}
+		if clix.Bool("enable") {
+			controllers, err := c.ListControllers()
+			if err != nil {
+				return err
+			}
+			if err := c.ToggleControllers(controllers, v2.Enable); err != nil {
```
Comment (AkihiroSuda) [inline cmd/cgctl/main.go]: This needs to be done for the parent group?

File: v2/paths.go [hunk 19]
```diff
@@ -19,42 +19,21 @@ package v2
...
 // This is commonly used for the Load function to restore an existing container
-func PidGroupPath(pid int) (GroupPath, error) {
+func PidGroupPath(pid int) (string, error) {
 	p := fmt.Sprintf("/proc/%d/cgroup", pid)
 	return parseCgroupFile(p)
 }
-
-// VerifyGroupPath verifies the format of g.
-// VerifyGroupPath doesn't verify whether g actually exists on the system.
-func VerifyGroupPath(g GroupPath) error {
```
Comment (AkihiroSuda) [inline v2/paths.go]: this function can be kept

File: cmd/cgctl/main.go (hunk 18 above)
Comment (crosbymichael) [reply cmd/cgctl/main.go]: you don't need to enable this at all if the system is setup to enable them

Comment (AkihiroSuda) [reply cmd/cgctl/main.go]: When you have `cpu` in the top group `/sys/fs/cgroup`, `cpu` is enabled for `/sys/fs/cgroup/foo`, but not enabled for `/sys/fs/cgroup/foo/bar`.

```console
root@suda-ws01:/sys/fs/cgroup# cat cgroup.controllers 
cpuset cpu io memory pids rdma
root@suda-ws01:/sys/fs/cgroup# mkdir foo
root@suda-ws01:/sys/fs/cgroup# cat foo/cgroup.controllers 
cpu io memory pids
root@suda-ws01:/sys/fs/cgroup# mkdir foo/bar
root@suda-ws01:/sys/fs/cgroup# cat foo/bar/cgroup.controllers 
root@suda-ws01:/sys/fs/cgroup#
```

File: v2/manager.go [hunk 20]
```diff
@@ -0,0 +1,310 @@
...
+}
+
+func (c *Manager) Stat() (map[string]uint64, error) {
+	controllers, err := c.ListControllers()
+	if err != nil {
+		return nil, err
+	}
+	out := make(map[string]uint64)
+	for _, controller := range controllers {
+		filename := fmt.Sprintf("%s.stat", controller)
```
Comment (AkihiroSuda) [inline v2/manager.go]: how will it work for other files such as `pids.current`?

Comment (AkihiroSuda) [reply v2/manager.go]: Probably we should either add file name to key or return `map[string]map[string]interface{}`

File: cmd/cgctl/main.go (hunk 18 above)
Comment (crosbymichael) [reply cmd/cgctl/main.go]: yes, its up to the creator of the cgroup to enable these.  I don't think the sub cgroup should modify the parent settings

File: v2/manager.go (hunk 20 above)
Comment (crosbymichael) [reply v2/manager.go]: i'ts still a uint

**Comment by crosbymichael** [comment]
We don't have to fix everything in this PR, it's meant to show that we can simplify things for v2 and to get this in so we can carry on the work

**Comment by AkihiroSuda** [comment]
follow-up: https://github.com/containerd/cgroups/pull/110

**Pull request merged**